src/lexer.coffee

```diff
@@ -803,6 +805,8 @@ exports.Lexer = class Lexer
     else if value in UNARY_MATH      then tag = 'UNARY_MATH'
     else if value in SHIFT           then tag = 'SHIFT'
     else if value is '?' and prev?.spaced then tag = 'BIN?'
+    else if value is '|' and (not prev?.spaced or not hasFollowingWhitespace) then tag = 'TYPE_ANNOTATION_DELIMITER'
```

Comment:
The `|` syntax for delimiting types and `::` for type signatures have the potential to clash with existing Coffeescript syntax: `|` for bitwise or and `::` as a shorthand for `prototype.`

But the idea was to distinguish them based on surrounding whitespace or lack thereof:

You'd have to always leave surrounding space on both sides of a bitwise or `|` (and correspondingly not leave space "inside" a type `|`)

And you'd have to leave space on both sides of a type signature `::` (and not try and do that with prototype shorthand `::`)